src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MatchXpathCheck.java

```diff
@@ -0,0 +1,285 @@
+ * |       |   |       `--IDENT -&gt; name [3:27]
+ * |       |   `--SEMI -&gt; ; [3:31]
+ * |       `--RCURLY -&gt; } [4:4]
+ * `--RCURLY -&gt; } [5:0]
+ * </pre>
+ * <p><b>-b</b> option shows AST nodes that match given Xpath query. This command can be used to
+ * validate accuracy of Xpath query against given file.</p>
+ * <pre>
+ * $ java -jar checkstyle-X.XX-all.jar Main.java -b "//METHOD_DEF[./IDENT[@text='sayHello']]"
+ * CLASS_DEF -&gt; CLASS_DEF [1:0]
+ * `--OBJBLOCK -&gt; OBJBLOCK [1:18]
+ * |--METHOD_DEF -&gt; METHOD_DEF [2:4]
+ * </pre>
+ * <p>
+ * The following example demonstrates validation of methods order, so that public methods should
+ * come before the private ones:
+ * </p>
+ * <pre>
+ * &lt;module name="MatchXpath"&gt;
+ *  &lt;property name="query" value="//METHOD_DEF[.//LITERAL_PRIVATE and
+ *  following-sibling::METHOD_DEF[.//LITERAL_PUBLIC]]"/&gt;
+ *  &lt;message key=&quot;matchxpath.match&quot;
+ *  value=&quot;Private methods must appear after public methods&quot;/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * public class Test {
+ *  public void method1() { }
+ *  private void method2() { } // violation
+ *  public void method3() { }
+ *  private void method4() { } // violation
+ *  public void method5() { }
+ *  private void method6() { } // ok
+ * }
+ * </pre>
+ * <p>
+ * To violate if there are any parametrized constructors
+ * </p>
+ * <pre>
+ * &lt;module name="MatchXpath"&gt;
+ *  &lt;property name="query" value="//CTOR_DEF[count(./PARAMETERS/*) &gt; 0]"/&gt;
+ *  &lt;message key=&quot;matchxpath.match&quot;
+ *  value=&quot;Parameterized constructors are not allowed&quot;/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * public class Test {
+ *  public Test(Object c) { } // violation
+ *  public Test(int a, HashMap&lt;String, Integer&gt; b) { } // violation
+ *  public Test() { } // ok
+ * }
+ * </pre>
+ * <p>
+ * To violate if method name is 'test' or 'foo'
+ * </p>
+ * <pre>
+ * &lt;module name="MatchXpath"&gt;
+ *  &lt;property name="query" value="//METHOD_DEF[./IDENT[@text='test' or @text='foo']]"/&gt;
+ *  &lt;message key=&quot;matchxpath.match&quot;
+ *  value=&quot;Method name should not be 'test' or 'foo'&quot;/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * public class Test {
+ *  public void test() {} // violation
+ *  public void getName() {} // ok
+ *  public void foo() {} // violation
+ *  public void sayHello() {} // ok
+ * }
+ * </pre>
+ * <p>
+ * To violate if new instance creation was done without <b>var</b> type
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;MatchXpath&quot;&gt;
+ *  &lt;property name=&quot;query&quot; value=&quot;//VARIABLE_DEF[./ASSIGN/EXPR/LITERAL_NEW
+ *  and not(./TYPE/IDENT[@text='var'])]&quot;/&gt;
+ *  &lt;message key=&quot;matchxpath.match&quot;
+ *  value=&quot;New instances should be created via 'var' keyword&quot;/&gt;
```

Comment:
`New instances should be created via 'var' keyword to avoid duplication of type reference in statement`

Reply:
done